tools/cldr-code/src/main/java/org/unicode/cldr/tool/LangTagsData.java

```diff
@@ -0,0 +1,387 @@
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.unicode.cldr.util.CLDRConfig;
+import org.unicode.cldr.util.CLDRFile;
+import org.unicode.cldr.util.CLDRLocale;
+import org.unicode.cldr.util.CLDRPaths;
+import org.unicode.cldr.util.CldrUtility;
+import org.unicode.cldr.util.Iso639Data;
+import org.unicode.cldr.util.Iso639Data.Type;
+import org.unicode.cldr.util.LanguageTagCanonicalizer;
+import org.unicode.cldr.util.LanguageTagParser;
+import org.unicode.cldr.util.StandardCodes.LstrType;
+import org.unicode.cldr.util.Validity;
+import org.unicode.cldr.util.Validity.Status;
+
+public class LangTagsData {
+    private final Pattern fullTagMatch = Pattern.compile("\\s*\"(full|tag)\": \"([^\"]+)\",");
+    private final String SIL = "sil1";
+
+    private final Splitter TAB_SPLITTER = Splitter.on('\t');
+    private final Set<String> LIKELY_SPECIALS = ImmutableSet.of("in", "iw", "ji", "jw", "mo");
+    private final Set<String> FIX_VALIDITY = ImmutableSet.of("Zanb");
+    private final Set<String> FIX_COUNTRY = ImmutableSet.of("yi");
+    private final Validity validity = Validity.getInstance();
+
+    private static final CLDRConfig CLDR_CONFIG = CLDRConfig.getInstance();
+    static final CLDRFile english = CLDR_CONFIG.getEnglish();
+
+    private static final LangTagsData INSTANCE = new LangTagsData();
+
+    private final Multimap<String, String> wikiData;
+    private final Map<String, LSRSource> jsonData;
+    private final Errors processErrors = new Errors();
+
+    private LangTagsData() {
+        wikiData = readWikidata();
+        jsonData = readJson();
+    }
+
+    public static LangTagsData getInstance() {
+        return INSTANCE;
+    }
+
+    public static Multimap<String, String> getWikiData() {
+        return getInstance().wikiData;
+    }
+
+    public static Map<String, LSRSource> getJsonData() {
+        return getInstance().jsonData;
+    }
+
+    public static Errors getProcessErrors() {
+        return getInstance().processErrors;
+    }
+
+    private Map<String, LSRSource> readJson() {
+
+        final LanguageTagParser ltpFull = new LanguageTagParser();
+        final LanguageTagParser ltpTag = new LanguageTagParser();
+
+        Path path = Paths.get(CLDRPaths.BIRTH_DATA_DIR, "/../external/langtags.json");
+        if (!Files.exists(path)) {
+            throw new IllegalArgumentException(path + " does not exist");
+        }
+
+        Matcher full = fullTagMatch.matcher("");
+        Map<LstrType, Status> errors = new TreeMap<>();
+
+        Output<String> lastFull = new Output<>();
+        Map<String, LSRSource> result = new TreeMap<>();
+        LanguageTagCanonicalizer langCanoner = new LanguageTagCanonicalizer(null);
+        try {
+            Files.lines(path)
```

Comment:
we shouldn't be parsing a JSON file using regex… (but this is preexisting code too).  Use `JsonStreamParser` (SAX) instead, or perhaps `gson.fromJson()` (DOM).

Reply:
Agreed. I didn't realize we had a JSON parser already in CLDR